cpp/ql/src/semmle/code/cpp/dataflow/internal/FlowVar.qll

```diff
@@ -133,17 +133,35 @@ private module PartialDefinitions {
   }
 
   /**
-   * A partial definition that's a definition by reference (in the sense of the
-   * `definitionByReference` predicate).
+   * A partial definition that's a definition by reference.
    */
   class DefinitionByReference extends PartialDefinition, TReferenceArgument {
-    VariableAccess va;
-
-    DefinitionByReference() { definitionByReference(va, definedExpr) }
-
-    VariableAccess getVariableAccess() { result = va }
+    Expr getArgument() { this = TReferenceArgument(result, _) }
+  }
 
-    override predicate partiallyDefines(Variable v) { va = v.getAnAccess() }
+  private predicate referenceArgument(Expr argument, VariableAccess va) {
+    argument = any(Call c).getAnArgument() and
+    exists(Type argumentType |
+      argumentType = argument.getFullyConverted().getType().stripTopLevelSpecifiers()
+    |
+      argumentType instanceof ReferenceType and
+      not argumentType.(ReferenceType).getBaseType().isConst() and
+      va = argument
+      or
+      argumentType instanceof PointerType and
+      not argumentType.(PointerType).getBaseType().isConst() and
+      (
+        // f(variable)
+        va = argument
+        or
+        // f(&variable)
+        va = argument.(AddressOfExpr).getOperand()
+        or
+        // f(&array[0])
```

Comment:
should this be generalized to handle multi-dimensional arrays?

Reply:
Is it normal to pass a multi-dimensional array as `f(&my3dArray[0][0][0])`? I'd expect people to just write `f(my3dArray)`. What should we do about `f(&my3dArray[0][0])` (two zeros)?

Can you find other common cases that aren't handled right here? I wouldn't be surprised if there are more problems.

Reply:
I don't think it would be normal to do that, but I wouldn't be surprised if something like `f(&my3dArray[x][y][z])` is significantly more common.

This implicitly covers `FieldAccess` as well - I assume that's intended?

Reply:
Looking at this again, I'm not at all sure if `FieldAccess` should be included here.

Reply:
Thank you for the reminder. I didn't see the `FieldAccess` comment in your last review.

The `referenceArgument` predicate defines the `DefinitionByReference` class, which is used for two things. The first thing is that it's [used in the `TBlockVar` constructor](https://github.com/Semmle/ql/blob/1bb57daf6fe659195e6323b8e9f8b188f948af33/cpp/ql/src/semmle/code/cpp/dataflow/internal/FlowVar.qll#L243), but that constructor requires `not v instanceof Field` and so excludes fields. The second thing is that it [determines the charpred of `DefinitionByReferenceNode`](https://github.com/Semmle/ql/blob/067c55adb948f9ef66150775eb24c159d7d9868b/cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowUtil.qll#L198). The `DefinitionByReferenceNode`s that don't assign to fields will have a `FlowVar` associated with them, and there will be flow out of them thanks to [this rule](https://github.com/Semmle/ql/blob/067c55adb948f9ef66150775eb24c159d7d9868b/cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowUtil.qll#L478-L480). Those that assign to fields will not have any flow out of them by default (because they have no `FlowVar`), but they will still be available in case users of the library want to add custom data flow edges out of them with `isAdditionalFlowStep`.

In summary, I think `FieldAccess` should be included here because it makes the library more flexible. You might even argue that arguments that have no `VariableAccess` should be included too, but then we're getting into corner cases that are probably best handled with IR-based data flow.